Add image enhancement measure (EME)

iliailmer opened the pull request:
## Description

This PR contains implementation of an image quality measure called EME, based on [this paper](https://www.semanticscholar.org/paper/A-New-Measure-of-Image-Enhancement-Agaian-Lentz/5adac5932775f089eaace7ebc45a6cba89809134).

It is based no the human visual system, as mentioned in the paper, and it can be used as a metric for enhancement algorithms to see if they work or not.

## Checklist

<!-- It's fine to submit PRs which are a work in progress! -->
<!-- But before they are merged, all PRs should provide: -->
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [x] Gallery example in `./doc/examples` (new features only)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [x] Unit tests
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

<!-- For detailed information on these and other aspects see -->
<!-- the scikit-image contribution guidelines. -->
<!-- https://scikit-image.org/docs/dev/contribute.html -->

## For reviewers

<!-- Don't remove the checklist below. -->
- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

iliailmer committed:
Proposition to add the following image quality assessment function

iliailmer committed:
Proposition to add the following image quality assessment function

iliailmer committed:
modified examples

pep8speaks commented:
Hello @iliailmer! Thanks for submitting the PR.

 - In the file [`skimage/image_quality_measure.py`](https://github.com/scikit-image/scikit-image/blob/130d9120000606975b7cbd4310bd633b45ee7b72/skimage/image_quality_measure.py), following are the PEP8 issues :

> [Line 42:80](https://github.com/scikit-image/scikit-image/blob/130d9120000606975b7cbd4310bd633b45ee7b72/skimage/image_quality_measure.py#L42): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 49:66](https://github.com/scikit-image/scikit-image/blob/130d9120000606975b7cbd4310bd633b45ee7b72/skimage/image_quality_measure.py#L49): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 50:66](https://github.com/scikit-image/scikit-image/blob/130d9120000606975b7cbd4310bd633b45ee7b72/skimage/image_quality_measure.py#L50): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 52:60](https://github.com/scikit-image/scikit-image/blob/130d9120000606975b7cbd4310bd633b45ee7b72/skimage/image_quality_measure.py#L52): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 53:60](https://github.com/scikit-image/scikit-image/blob/130d9120000606975b7cbd4310bd633b45ee7b72/skimage/image_quality_measure.py#L53): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 54:21](https://github.com/scikit-image/scikit-image/blob/130d9120000606975b7cbd4310bd633b45ee7b72/skimage/image_quality_measure.py#L54): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator

iliailmer committed:
pep8 corrections

iliailmer committed:
pep8 corrections

grlee77 commented:
Hi @iliailmer,

Thank you for the contribution! I have not yet done a detailed review of the publication, but it has a substantial number of citations and the implementation looks simple and should be easy to maintain going forward. I think this is a reasonable candidate for inclusion in scikit-image.

I can do a more thorough review over the coming week, but a few initial comments:

1.) The function should not be stored in the top-level `skimage` directory.  A more appropriate location would be in the `metrics` submodule, perhaps as an addition to the current [simple_metrics.py](https://github.com/scikit-image/scikit-image/blob/master/skimage/measure/simple_metrics.py)

2.) Use a more descriptive name for the function than `quality`. Perhaps `enhancement`, `enhancement_measure` or something similarly descriptive. For example the existing functions `compare_psnr` and `compare_ssim` are measures of image similarity often referred to as image quality metrics. Having PSNR or SSIM in the function name gives some immediate indication of what "quality" is being measured.

3.) It looks like this can be trivially extended to the n-dimensional case rather than just 2d and 3d. We can help with this if you have questions on how to approach it. A convention in the library is to include a `multichannel` argument so that the user can indicate whether the final axis of an array is a "spatial" axis or whether it represents channels (e.g R, G, B channels for a color image, for example).

iliailmer committed:
changed name to 'enhancement_measure', placed the funciton into measure/simple_metrics.py, added footprint parameter for n-dimensional filtering

iliailmer commented:
Hi @grlee77 , thanks for your comments!

Regarding the multidimensional extension, I realize now that the max/min filters used here are, by default, usable with any number of dimensions, so one does not need to check the length of the array. In this case the `multidimensional` parameter may not be required, and the filter shape (`footprint`) will be inferred automatically.

iliailmer committed:
some pep8 corrections

lagru commented on skimage/image_quality_measure.py:
```diff
@@ -0,0 +1,53 @@
+        >>> print(f"\tbefore: {quality(img)}")
+                before: 0.9096745071475523
+        >>> print(f"\tafter: {quality(equalize_hist(img))}")
+                after: 1.299327371881219
```
The docstring formatting is a little unconventional which may be the reason for the doctest failure on travis. What is you reasoning for using print functions? I think the indentation after the print statements confuses the parser. You could use comments instead to annotate "before" and "after". E.g.:

```python
>>> quality(img)  # before
0.9096745071475523
>>> quality(equalize_hist(img))  # after
1.299327371881219
```

And just a nitpick: could you remove one indentation level from the docstring? It's not really necessary.

lagru replied:
Oh, I commented in the wrong file. But it applies to the other one as well.

iliailmer replied:
Thanks for the comments @lagru, I did not really think about why I put the print statements there. Makes sense to remove them in this contexts, true.

lagru commented:
Looks promising! I'll have a more thorough look later.

lagru commented on skimage/measure/simple_metrics.py:
```diff
@@ -142,3 +145,50 @@ def compare_psnr(im_true, im_test, data_range=None):
+    """
+    image = img_as_float(image)
+    eme = np.zeros_like(image)
+    footprint = tuple([size] * len(image.shape))
```
You don't need to construct the `footprint` here. You can just pass `size` directly to the filter functions as an argument of the same name.

lagru commented on skimage/measure/simple_metrics.py:
```diff
@@ -142,3 +145,50 @@ def compare_psnr(im_true, im_test, data_range=None):
+    eme = np.zeros_like(image)
+    footprint = tuple([size] * len(image.shape))
+    eme = np.divide(maximum_filter(image, footprint=footprint) + 1,
+                    minimum_filter(image, footprint=footprint) + 1)
```
Is adding `+1` here supported by the linked paper? Why is this necessary? Do you want to avoid division by 0 this way?

iliailmer replied:
So, in the paper the authors don't add anything, but division by zero is a possibility here, so to avoid it I add +1 to both numerator and denominator. At least in my experiments I used the function in this form a lot. It might make sense to use something like an `eps = 1` as a parameter passed to the function. I will check which values make sense.
Should the implementation follow the same comparison pattern as with PSNE and other functions? `compare_enhancement` or something like that?

iliailmer replied:
Sorry, forgot to address the question about the image. I don't personally have it, but google search gave me [this link](https://link.springer.com/chapter/10.1007%2F978-3-642-23154-4_11) to the source paper and [the image](https://www.google.com/search?sa=G&hl=en&q=monochrome&tbm=isch&source=iu&ictx=1&tbs=simg:CAESvgIJt1bAc_1cJHhMasgILEKjU2AQaBAgUCAoMCxCwjKcIGmEKXwgDEieVAYESlwH6AYMS7AV_1qgbMAvsBiz7wJ6c27if2NaU29DbyJ6I2kz4aMBrLMEAUP2pMseBtDd5bfccLZkF_1HpcsD1s4kW6GfqcgdI3SV3g-nobzAkj7i1DbYSAEDAsQjq7-CBoKCggIARIE6a_1fCgwLEJ3twQkangEKHQoKbW9ub2Nocm9tZdqliPYDCwoJL20vMDFtd2tmCh4KC2FsYXJtIGNsb2Nr2qWI9gMLCgkvbS8wNDZkbHIKGgoHYW50aXF1ZdqliPYDCwoJL20vMDF5czJzCiIKD2JsYWNrLWFuZC13aGl0ZdqliPYDCwoJL20vMDFnNmdzCh0KCnN0aWxsIGxpZmXapYj2AwsKCS9tLzAxc3B6cww&fir=VQ-TGHm2qMMWYM%253A%252CvLAo2pnVeI90MM%252C_&vet=1&usg=AI4_-kQx9DbbxFxkhS3ywjL0w-seNQZmRQ&ved=2ahUKEwjR3435g_TgAhUJVN8KHRurCmoQ9QEwAnoECAAQBA#imgrc=VQ-TGHm2qMMWYM:). Should the image be added to tests here?

lagru replied:
I'm not sure that simply adding +1 is the right way to go. According to [this guide](http://scikit-image.org/docs/stable/user_guide/data_types.html) images of dtype float may also range from -1 to 1. If we pass in an image with a -1 somewhere adding +1 doesn't work.

No, adding the image to the tests shouldn't be necessary. If you want to use a natural image there you can use the [ones already shipped with scikit-image](http://scikit-image.org/docs/stable/api/skimage.data.html).

lagru reviewed (commented):
Thanks for addressing @grlee77's feedback. Good work so far! With the recent changes this should work for the n-dimensional case, perhaps we could mention that in the docstring. I think you forgot to remove the now obsolete file `image_quality_measure`.  Could you fix that?

Otherwise I don't yet fully understand your implementation when looking at the linked paper (see comment). Do you or someone else know a source for the clock image used in the paper? It would be useful to compare results of this function to the paper.

iliailmer committed:
change to examples

iliailmer committed:
added name to __all__, fixed examples"

lagru commented:
@iliailmer 

> I think you forgot to remove the now obsolete file image_quality_measure. Could you fix that?

I'm getting confused now. The diff on GitHub shows that this PR still includes 2 redundant files with nearly the same code.  Could you ensure that the redundant files are removed? We only want to keep the changes in `skimage/measure/simple_metrics.py` as suggested by @grlee77.

lagru commented on skimage/measure/simple_metrics.py:
```diff
@@ -142,3 +146,47 @@ def compare_psnr(im_true, im_test, data_range=None):
+    image = img_as_float(image)
+    eme = np.zeros_like(image)
+    eme = np.divide(maximum_filter(image, tuple([size]*len(image.shape))) + 1,
+                    minimum_filter(image, tuple([size]*len(image.shape))) + 1)
```
Sorry, what I meant earlier is that [`maximum_filter`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.maximum_filter.html) already supports the parameter `size` directly. E.g.

```python
eme = np.divide(maximum_filter(image, size=size)) + 1,
                minimum_filter(image, size=size)) + 1)
```
Or am I reading that wrong?

iliailmer replied:
I think you're right. I was using the `footprint` parameter, because I got the impression from the docs that it is necessary for proper shape of the filter. I tested with just using `size' and it seems to work as well.

iliailmer committed:
removed unnecessary files

iliailmer committed:
filter parameter  is used instead of

iliailmer committed:
added epsilon instead of +1 to avoid division by zero.

iliailmer committed:
added intensity rescaling to account for images with negative intensities

lagru commented on skimage/measure/simple_metrics.py:
```diff
@@ -142,3 +147,51 @@ def compare_psnr(im_true, im_test, data_range=None):
+        >>> from skimage.data import camera
+        >>> from skimage.exposure import equalize_hist
+        >>> img = camera()
+        >>> quality(img) # before
```
`quality` is no longer the correct function name which is the reason for the failures on travis.

lagru commented on skimage/measure/simple_metrics.py:
```diff
@@ -142,3 +147,51 @@ def compare_psnr(im_true, im_test, data_range=None):
+
+    image = img_as_float(image)
+    image = rescale_intensity(image, out_range=(0., 1.))
+    eme = np.zeros_like(image)
```
I think this line is not needed.

iliailmer replied:
> Is there a reason to rescale between [0, 1] in particular? If we were to chose a value above zero as the lower bound we could omit the `eps` parameter altogether.

Thanks for the feedback, I will get to all the stuff mentioned above as soon as I can.
Meanwhile, regarding rescaling, when I was testing images with range `[-1,1]` I ran into issues with logarithms, since some pixel neighborhoods can have multiple negative values, the logarithm would try to evaluate negatives and return `nan`. Hence the rescaling to `[0,1]`.

lagru reviewed (commented):
@iliailmer thanks for addressing my feedback. The solution with `eps` as an optional parameter is much better now but I'm not sure whether this is the right way to solve this. As I understand the paper the formula is meant to quantify a contrast improvement somehow. While rescaling the image won't affect the [Weber contrast](https://en.wikipedia.org/wiki/Contrast_(vision)#Weber_contrast) I'm unsure about this measurement. 

Is there a reason to rescale between [0, 1] in particular? If we were to chose a value above zero as the lower bound we could omit the `eps` parameter altogether.

iliailmer committed:
doctest fixes

emmanuelle commented:
Hi @iliailmer thanks for your PR. It looks good, but we're missing unit tests (write unit tests on small images, 2D and 3D, with int and float types for example) as well as a gallery example.

iliailmer commented:
@emmanuelle Thank you, I will get to it as soon as I can. Should the test images come from the `data` module of scikit-image?

emmanuelle commented:
@iliailmer yes, either images from ``skimage.data`` (you can downscale them for faster execution, or choose a ROI, since we want tests to execute fast enough) or synthetic data which you create yourself. You can take a look at the different ``tests`` folders in each submodule of scikit-image, to get an idea of how tests are written.

Commit:
added unit tests and fixed some numerial stability issues

Commit:
some identation

emmanuelle commented on skimage/measure/simple_metrics.py:
```diff
@@ -142,3 +145,48 @@ def compare_psnr(im_true, im_test, data_range=None):
+            Can be either 3-channel RGB or 1-channel grayscale.
+        size : int
+            Size of the window.
+        eps : float
```
Do you need ``eps`` as a parameter or could you choose a value inside the function once and for all?

iliailmer replied:
I think it is better to let the user control the `eps` parameter because if the potential image/matrix is sparse. A larger `eps` could help to avoid unnecessary negative values.

emmanuelle commented on skimage/measure/simple_metrics.py:
```diff
@@ -142,3 +145,48 @@ def compare_psnr(im_true, im_test, data_range=None):
+def enhancement_measure(image: np.ndarray,
+                        eps: float = 1e-6,
+                        size: int = 3) -> float:
+    """ The image enhancement measure called EME based on [1]_.
```
Add one sentence to explain what EME is measuring, like "EME is a measure of average local contrast, measured on a logarithmic scale inside windows of user-provided size".

iliailmer replied:
I added more explanations to the doc files, thanks for pointing that out.

emmanuelle commented:
Thanks for the update! I left two more minor comments, and we would also need a gallery example :-).

iliailmer committed:
added explanations in doc

iliailmer committed:
gallery example for enhancement measure

emmanuelle commented on doc/examples/filters/plot_enhancemet_measure.py:
```diff
@@ -0,0 +1,65 @@
+"""
```
There is a typo in the file name enhancemet --> enhancement

emmanuelle commented on doc/examples/filters/plot_enhancemet_measure.py:
```diff
@@ -0,0 +1,65 @@
+
+import numpy as np
+import matplotlib
+matplotlib.use("TkAgg")
```
These two lines above are not needed and should be removed.

iliailmer replied:
Thank you for your comments! I fixed the issues.

emmanuelle reviewed (commented):
Thanks, we're really getting there. Two more very minor things to correct in the example; plus there is some weird doc build failure related to the example, I'll have to investigate more to understand what's going on. @scikit-image/core we'll soon need a second review.

sciunto commented on skimage/measure/simple_metrics.py:
```diff
@@ -142,3 +145,52 @@ def compare_psnr(im_true, im_test, data_range=None):
+        image : array
+            Input image of which the quality should be assessed.
+            Can be either 3-channel RGB or 1-channel grayscale.
+        size : int
```
```suggestion
        size : int, optional
```

sciunto commented on skimage/measure/simple_metrics.py:
```diff
@@ -142,3 +145,52 @@ def compare_psnr(im_true, im_test, data_range=None):
+            Can be either 3-channel RGB or 1-channel grayscale.
+        size : int
+            Size of the window.
+        eps : float
```
```suggestion
        eps : float, optional
```

sciunto commented on skimage/measure/simple_metrics.py:
```diff
@@ -142,3 +145,52 @@ def compare_psnr(im_true, im_test, data_range=None):
+
+def enhancement_measure(image: np.ndarray,
+                        eps: float = 1e-6,
+                        size: int = 3) -> float:
```
Please, keep a consistent arg order in the signature and the docstring.

iliailmer replied:
@sciunto Thank you for pointing that out, I fixed it.

sciunto commented on skimage/measure/simple_metrics.py:
```diff
@@ -142,3 +145,52 @@ def compare_psnr(im_true, im_test, data_range=None):
+        .. [1] Agaian, Sos S., Karen Panetta, and Artyom M. Grigoryan.
+               "A new measure of image enhancement."
+               IASTED International Conference on Signal Processing
+               & Communication. Citeseer, 2000.
```
Please, add a DOI with the format (auto formatting):

:DOI:`xxx`

sciunto commented on doc/examples/filters/plot_enhancemet_measure.py:
```diff
@@ -0,0 +1,65 @@
+# .. [1] Agaian, Sos S., Karen Panetta, and Artyom M. Grigoryan.
+# "A new measure of image enhancement."
+# IASTED International Conference on Signal Processing
+# & Communication. Citeseer, 2000.
```
Please, add the DOI here as well.

sciunto committed:
Update skimage/measure/simple_metrics.py

Co-Authored-By: iliailmer <i.ilmer@icloud.com>

iliailmer committed:
added DOI, fixed name typo, improved consistency in parameter order

stefanv commented on doc/examples/filters/plot_enhancement_measure.py:
```diff
@@ -0,0 +1,64 @@
+Quantitative Measure of Image Enhancement
+=========================================
+
+In this example, we will measure the result of applying contrast enhancement
```
```suggestion
In this example, we will quantify the effect of applying contrast enhancement
```

stefanv commented on doc/examples/filters/plot_enhancement_measure.py:
```diff
@@ -0,0 +1,64 @@
+=========================================
+
+In this example, we will measure the result of applying contrast enhancement
+to the image. We will use a measure called EME [1]_.
```
```suggestion
to an image using a measure called EME [1]_.
```

stefanv commented on doc/examples/filters/plot_enhancement_measure.py:
```diff
@@ -0,0 +1,64 @@
+In this example, we will measure the result of applying contrast enhancement
+to the image. We will use a measure called EME [1]_.
+
+It is defined as a log of the ratio of local maximum to local minimum
```
```suggestion
It is defined as the log of the ratio of local maximum to minimum
```

stefanv commented on doc/examples/filters/plot_enhancement_measure.py:
```diff
@@ -0,0 +1,64 @@
+to the image. We will use a measure called EME [1]_.
+
+It is defined as a log of the ratio of local maximum to local minimum
+within a sliding window. The code allows the user to choose the window size.
```
```suggestion
within a sliding window of given size.
```

stefanv commented on doc/examples/filters/plot_enhancement_measure.py:
```diff
@@ -0,0 +1,64 @@
+
+It is defined as a log of the ratio of local maximum to local minimum
+within a sliding window. The code allows the user to choose the window size.
+This function was defined to quantify improvement of the image after processing.
```
```suggestion
This measure aims to provide a perceptual estimate of image quality, useful when comparing image enhancement algorithms.
```

stefanv commented on doc/examples/filters/plot_enhancement_measure.py:
```diff
@@ -0,0 +1,64 @@
+from skimage import data
+from skimage.filters import gaussian
+from skimage.exposure import equalize_adapthist
+from skimage.measure.simple_metrics import enhancement_measure
```
We don't typically do nested level imports in scikit-image.  Please import directly from `skimage.measure`.

stefanv commented on doc/examples/filters/plot_enhancement_measure.py:
```diff
@@ -0,0 +1,64 @@
+img_rgb = data.astronaut()
+
+
+def plotter(image_1: np.ndarray, image_2: np.ndarray):
```
We need a better name than `plotter` here: `compare_images` or `side_by_side_display` or something.

stefanv commented on doc/examples/filters/plot_enhancement_measure.py:
```diff
@@ -0,0 +1,64 @@
+    after = transform(before, **kwargs)
+    plotter(before, after)
+
+compare_eme(before=img_gray, transform=gaussian, kwargs={'sigma':10})
```
This is only run twice, so probably overkill.  How about simply:

```
compare(image, gaussian(image, sigma=10))
compare(image, equalize_adapthist(image, nbins=256 * 3))
```

stefanv commented on doc/examples/filters/plot_enhancement_measure.py:
```diff
@@ -0,0 +1,64 @@
+# References
+# ----------
+# .. [1] Agaian, Sos S., Karen Panetta, and Artyom M. Grigoryan.
+# "A new measure of image enhancement."
```
This reference does not seem to be correctly formatted; needs some indentation?

stefanv commented on skimage/measure/simple_metrics.py:
```diff
@@ -142,3 +145,53 @@ def compare_psnr(im_true, im_test, data_range=None):
+        >>> from skimage.data import camera
+        >>> from skimage.exposure import equalize_hist
+        >>> img = camera()
+        >>> before = enhancement_measure(img) # before
```
```suggestion
        >>> before = enhancement_measure(img)  # before
```

stefanv commented on skimage/measure/tests/test_simple_metrics.py:
```diff
@@ -75,3 +74,23 @@ def test_NRMSE_errors():
         compare_nrmse(x, x, 'foo')
+
+
+def test_EME_greyscale():
```
Does the original paper provide any absolute values for given images that we can try and reproduce?  Only checking whether the value goes down after enhancement is not quite strict enough.

lagru replied:
This came up earlier (see [here](https://github.com/scikit-image/scikit-image/pull/3776#discussion_r263983133)). The image in question could be a good candidate for comparison. Unfortunately its source isn't mentioned inside the paper and the one found through searching may not be *exactly* the same.

Another issue is that the absolute value may change depending on `eps`. That parameter can help to avoid division by zero but is an addition of this PR and not part of the paper.

lagru replied:
Furthermore the rescaling between 0 and 1 isn't part of the paper either, so I doubt we'd get comparable absolute values even if we used the same image as in the paper.

stefanv replied:
That's fine; I was hoping for some rough agreement between their work and ours, otherwise how do we know we're even in the right ballpark?  Overall, any test that we can define that is a bit more strict will be helpful.

iliailmer replied:
@stefanv @lagru I looked at the experiment section of the paper, I managed to find the moon surface image [here](https://homepages.inf.ed.ac.uk/rbf/HIPR2/images/moo1.gif), running the code with that image without converting to float and (0,1)-rescaling, I can get the value to be ~6.67 if the `size` is 6 and `eps` is `1e-8`. The same parameters yield ~6.28 when converted to float and rescaled.

stefanv replied:
Does this correspond to the values in the paper?

iliailmer replied:
Yes, the paper lists 6.78 on page 4, Figure 1b.

stefanv replied:
OK, so we can tighten up one of the tests to match against a specific value.

Why are the values different depending on the dtype and scaling?

iliailmer replied:
I need to think some more about it, right now I am not sure why this is the case.

stefanv reviewed (changes requested):
Thanks for your contribution!  I've added a few small suggestions.

stefanv committed:
Update skimage/measure/simple_metrics.py

Co-Authored-By: iliailmer <i.ilmer@icloud.com>

stefanv committed:
Update doc/examples/filters/plot_enhancement_measure.py

Co-Authored-By: iliailmer <i.ilmer@icloud.com>

stefanv committed:
Update doc/examples/filters/plot_enhancement_measure.py

Co-Authored-By: iliailmer <i.ilmer@icloud.com>

stefanv committed:
Update doc/examples/filters/plot_enhancement_measure.py

Co-Authored-By: iliailmer <i.ilmer@icloud.com>

stefanv committed:
Update doc/examples/filters/plot_enhancement_measure.py

Co-Authored-By: iliailmer <i.ilmer@icloud.com>

iliailmer committed:
doc file corrections, reference reformatting

iliailmer committed:
doc example corrections

iliailmer committed:
doc example corrections

stefanv committed:
Update doc/examples/filters/plot_enhancement_measure.py

Co-Authored-By: iliailmer <i.ilmer@icloud.com>

iliailmer committed:
small changes to citation format

iliailmer committed:
fixed critical error in example

iliailmer committed:
plt.show -> fig.show

stefanv commented on doc/examples/filters/plot_enhancement_measure.py:
```diff
@@ -0,0 +1,69 @@
+to an image using a measure called EME [1]_.
+
+It is defined as the log of the ratio of local maximum to minimum
+<<<<<<< HEAD
```
woops.

stefanv commented on doc/examples/filters/plot_enhancement_measure.py:
```diff
@@ -0,0 +1,69 @@
+
+It is defined as the log of the ratio of local maximum to minimum
+<<<<<<< HEAD
+within a sliding window. The code allows the user to choose the window size.
```
```suggestion
within a sliding window of given size.
```

stefanv commented on doc/examples/filters/plot_enhancement_measure.py:
```diff
@@ -0,0 +1,69 @@
+within a sliding window. The code allows the user to choose the window size.
+This function was defined to quantify improvement of the image
+after processing.
+=======
```
Another artifact.

stefanv commented on doc/examples/filters/plot_enhancement_measure.py:
```diff
@@ -0,0 +1,69 @@
+It is defined as the log of the ratio of local maximum to minimum
+<<<<<<< HEAD
+within a sliding window. The code allows the user to choose the window size.
+This function was defined to quantify improvement of the image
```
How about: This method was developed to quantify improvement of the image after processing.

stefanv commented on skimage/measure/simple_metrics.py:
```diff
@@ -142,3 +145,53 @@ def compare_psnr(im_true, im_test, data_range=None):
+        >>> from skimage.data import camera
+        >>> from skimage.exposure import equalize_hist
+        >>> img = camera()
+        >>> before = enhancement_measure(img)  # before
```
```suggestion
        >>> before = enhancement_measure(img)
```

iliailmer committed:
fixed artifacts in docs and comments

iliailmer committed:
added blank line to docs

lagru commented on doc/examples/filters/plot_enhancement_measure.py:
```diff
@@ -0,0 +1,61 @@
+# References
+# ----------
+# .. [1] Sos S. Agaian, Karen Panetta, and Artyom M. Grigoryan.
+# "A new measure of image enhancement.",
```
I'm guessing that the failures on Travis are due to the missing indent here and the three lines below. Adding 3 spaces should fix this.

iliailmer replied:
I added the corrections, thanks for pointing that out

iliailmer committed:
Merge pull request #1 from scikit-image/master

fork update

sciunto commented:
I mark this PR to 0.17 as a targeted milestone to get time for finishing this PR. Some conflicts are due to a recent code refactoring.

iliailmer committed:
edited formatting for the example

iliailmer committed:
Merge branch 'master' into image-quality

iliailmer committed:
PEP8 corrections

iliailmer committed:
Merge remote-tracking branch 'upstream/master'

iliailmer committed:
Merge remote-tracking branch 'upstream/master'

iliailmer committed:
fixed conflict in test_simple_metrics.py

iliailmer committed:
added '_' to name 'correct_mesh_orientation'

iliailmer committed:
fixed deprecation warnings

iliailmer committed:
remove unnecessary function call, changed args to call for normalized_root_mse

iliailmer committed:
one more fix for normalized_root_mse

iliailmer committed:
pep8 fix

iliailmer committed:
docstring fix: blank line

iliailmer committed:
docstring fix: blank line

iliailmer committed:
docstring fix: blank line

iliailmer committed:
docstring fix: blank line

iliailmer commented:
I keep getting 
`Warning, treated as error:
C:\hostedtoolcache\windows\Python\3.6.8\x64\lib\site-packages\skimage\measure\simple_metrics.py:docstring of skimage.measure.compare_mse:23:Explicit markup ends without a blank line; unexpected unindent.
`
`
mingw32-make[1]: *** [Makefile:53: html] Error 2
`
`
mingw32-make[1]: Leaving directory 'd:/a/1/s/doc'
`
`
mingw32-make: *** [Makefile:27: html] Error 2
`

Tried to fix docstring couple of times already, not sure where the unindent is happening

rfezzani commented:
>Tried to fix docstring couple of times already, not sure where the unindent is happening

Have you tried the solution proposed [here](https://stackoverflow.com/questions/38761546/markup-ending-without-empty-line-but-markup-is-only-thing-in-comment)?

iliailmer committed:
docstring fix

iliailmer commented:
@rfezzani Thank you, it worked.

sciunto commented on skimage/measure/simple_metrics.py:
```diff
@@ -71,12 +78,63 @@ def compare_psnr(im_true, im_test, data_range=None):
+            "A new measure of image enhancement.",
+            IASTED International Conference on Signal Processing
+            & Communication, Citeseer, 2000,
+            :DOI:10.1.1.35.4021
```
```suggestion
            :DOI:`10.1.1.35.4021`
```

sciunto replied:
The DOI is incorrect. Also, backquotes must be set as suggested.

iliailmer replied:
Hi @sciunto, thanks for pointing that our!
I was having troubles with the DOI though.

The only source that lists the DOI is [this](https://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.35.4021) link, and I doubt that there is a better version or at least i couldn't find one. The [doi.org](doi.org) says this is only a prefix, but google search for doi=10.1.1.35.4021 returns the link above no problem. The original journal is not available (see [here](https://www.iasted.org/CONFERENCES/pastinfo-320.html)) so I am not sure what is the best way to prcoceed here.

sciunto replied:
I think doi.org must be the reference. If no working doi can be found, it is better to do not display a doi. Eventually, a link to the pdf file can be added instead.

iliailmer replied:
I removed it for now.

sciunto commented on doc/examples/filters/plot_enhancement_measure.py:
```diff
@@ -0,0 +1,61 @@
+#        "A new measure of image enhancement.",
+#        IASTED International Conference on Signal Processing
+#        & Communication, Citeseer, 2000,
+#        :DOI:10.1.1.35.4021
```
Same here.

iliailmer committed:
doi backquotes

iliailmer committed:
removed doi

iliailmer commented:
> Thanks for your contribution! I've added a few small suggestions.

Hi @stefanv, could you please take a look and check if anything is amiss? All the checks are passed so far. Thanks!

lagru commented:
Hey @iliailmer, sorry for the slow review process. I still want to get this into scikit-image! There are some merge conflicts now, do you want to resolve these or shall I? Afterwards I can do another review and hopefully approve to get this going again. :pray:

lagru committed:
Merge branch 'main' into pr3776-image-quality

lagru committed:
Re-add enhancement_measure after merge

simple_metrics was moved from skimage/measure to skimage/metrics making
it quite complicated to preserve the changes during the merge. This
commit re-adds the original changes authored by iliailmer
<i.ilmer@icloud.com> back.

Co-authored-by: iliailmer <i.ilmer@icloud.com>

lagru commented:
Hey, I took the liberty to merge this which was a little complicated due to us moving `simple_metrics` from `skimage/measure` to `skimage/metrics` on `main`. I hope I got every relevant change pertaining to adding the new function `enhancement_measure`.

lagru commented on skimage/metrics/simple_metrics.py:
```diff
@@ -259,3 +263,53 @@ def normalized_mutual_information(image0, image1, *, bins=100):
+    >>> before = enhancement_measure(img)
+    >>> after = enhancement_measure(equalize_hist(img))
+    >>> before < after
+    True
```
Other measure functions in this module are applied on and compare two images. As far as I understand, the EME of an image is not useful on its own but rather it's difference to the EME of an improved image. It might be worth thinking about making the function take two images and return a bool or difference between the EME of both images. That way we also prevent users from assigning to much meaning to the absolute value of the measure.

lagru commented on doc/examples/filters/plot_enhancement_measure.py:
```diff
@@ -0,0 +1,61 @@
+estimate of image quality, useful when comparing image enhancement algorithms.
+
+"""
+import numpy as np
```
```suggestion
```
Not used.

lagru commented on doc/examples/filters/plot_enhancement_measure.py:
```diff
@@ -0,0 +1,61 @@
+
+
+#####################################################################
+# We will create a function that takes an original image as `before`
+# and applies a `transform` to it. We will use a gaussian filter to
```
Nitpick: the function doesn't actually apply a transform to it. It just calls `enhancement_measure` on both and plots the results. Rewording this might be a lot clearer to new users.

lagru committed:
Remove type hints and improve docstring

Type hints are not (yet) used within scikit-image. Within the docstring
I tried to describe the measure in more precise terms. However, the
paper seems a bit vague on the interpretation of this measure.

lagru committed:
Add a link to the reference paper

lagru commented on skimage/metrics/simple_metrics.py:
```diff
@@ -259,3 +263,51 @@ def normalized_mutual_information(image0, image1, *, bins=100):
+    ----------
+    image : array
+        Input image of which the quality should be assessed.
+        Can be either 3-channel RGB or 1-channel grayscale.
```
We should probably support only grayscale images as input to make sure that the user determines [how the total luminance of a colored image is calculated](https://scikit-image.org/docs/stable/api/skimage.color.html#skimage.color.rgb2gray). Currently, the way that `maximum_filter` and `minimum_filter` is combining color channels is not something that is covered by the paper.

lagru commented on skimage/metrics/simple_metrics.py:
```diff
@@ -259,3 +263,51 @@ def normalized_mutual_information(image0, image1, *, bins=100):
     return (H0 + H1) / H01
+
+
+def enhancement_measure(image, size=3, eps=1e-6):
```
Is there some reasoning between choosing `eps=1e-6` as the default? Is something else more reasonable?

lagru commented on doc/examples/filters/plot_enhancement_measure.py:
```diff
@@ -0,0 +1,60 @@
+
+
+def compare_side_by_side(before, after):
+    fig, ax = plt.subplots(nrows=1, ncols=2, figsize=(16, 14))
```
```suggestion
    fig, ax = plt.subplots(nrows=1, ncols=2)
```
I think removing the forced figure size leads to a nicer layout. Otherwise the images seem very small.

lagru commented on skimage/metrics/tests/test_simple_metrics.py:
```diff
@@ -137,3 +140,23 @@ def test_nmi_random_3d():
+    assert_less(enhancement_measure(orig, size=5),
+                enhancement_measure(enhanced, size=5))
+    assert_less(enhancement_measure(orig, size=11),
+                enhancement_measure(enhanced, size=11))
```
We should probably check the actual values of the EMEs and that they are within an expected range. That way our tests would inform us if some change leads to significantly different EMEs.

lagru reviewed (commented):
I took the liberty and tried to make the docstring use a more precise description of the EME. However, I'm still a bit vague on how to EME should be interpreted precisely and what kind of "enhancement" it measures. Does it purely measure contrast enhancement? To me that seems to me where we can gain the most value yet.

Otherwise I suggested some smaller but easier changes that should probably be addressed.

lagru commented:
Hey @iliailmer, would you be able to look into the feedback? This seems like something that would be cool to have in scikit-image. It's also totally fine if you don't have the bandwidth in which case maybe someone else could take over. :)

matthew-brett commented:
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration` branch.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

In addition - our automated script detected that your branch had merge conflicts with the pre-Megamove branch. Please make sure to merge in the `pre-megamove-integration` branch, and resolve the merge conflicts in the way that you want.

Read on for the Megamove rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
